examples/asr/asr_adapters/train_asr_adapter.py

```diff
@@ -0,0 +1,146 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+# Adapting the model
+
+python train_asr_adapter.py \
+    --config-path="conf/" \
+    --config-name="asr_adaptation.yaml" \
+    model.pretrained_model=null \
+    model.nemo_model=null \
+    model.adapter.adapter_name=<Unique adapter name> \
+    model.adapter.in_features=<dimension of the layer outputs of the model> \
+    model.adapter.dim=32 \
+    model.train_ds.manifest_filepath=<Path to manifest> \
+    model.train_ds.batch_size=16 \
+    model.validation_ds.manifest_filepath=<Path to manifest> \
+    model.validation_ds.batch_size=16 \
+    model.optim.lr=0.5 \
+    model.optim.weight_decay=0.0 \
+    model.optim.sched.warmup_steps=100 \
+    trainer.max_steps=300 \
+    trainer.devices=1 \
+    trainer.precision=32 \
+    exp_manager.exp_dir=<Some directory for experiment manager>
+
+# Fine-tune a model
+
+While adaptation is very efficient for low-resource datasets, it imposes several restrictions -
+
+- The vocabulary of the new dataset must be supported by the pre-existing vocabulary or tokenizer.
+    If tokens exist outside this scope, the adapter will have to learn UNK tokens (or fail entirely
+    for character based models).
+
+- As a consequence of the above, the language of the new dataset must be the same as the original model.
+    There is ongoing research to enable more sophisticated adapters for other languages.
+
+When adapters cannot be readily used due to the above limitations, fine-tuning may be a better alternative.
+
+For documentation on fine-tuning a model, please visit -
+https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/configs.html#fine-tuning-configurations
+
+# Pretrained Models
+
+For documentation on existing pretrained models, please visit -
+https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/results.html
+
+"""
+
+import pytorch_lightning as pl
+from omegaconf import OmegaConf, open_dict
+
+from nemo.collections.asr.models import ASRModel
+from nemo.core.config import hydra_runner
+from nemo.utils import logging
+from nemo.utils.exp_manager import exp_manager
+
+
+def update_encoder_config_to_support_adapter(model_cfg, new_cfg):
+    with open_dict(model_cfg):
+        if 'Adapter' not in model_cfg.encoder._target_:
+            model_cfg.encoder._target_ = model_cfg.encoder._target_ + 'Adapter'
+
+
+def update_model_cfg(original_cfg, new_cfg):
+    with open_dict(new_cfg):
+        # drop keys which dont exist in old config
+        new_keys = list(new_cfg.keys())
+        for key in new_keys:
+            if key not in original_cfg:
+                new_cfg.pop(key)
+                print("Removing unavailable key from config :", key)
```

Comment:
nit: do we have conformity of only using f-string formatting syntax in NeMo? I saw mixed use cases.